Documenting type class options, typeclasses eauto

mattam82 opened the pull request:
This PR is actually twofold, as I found out that some piece of syntax was missing to really access the new features in 8.6 while documenting it. So one commit is about documenting all the options of typeclass resolution (and deprecating one, Set Typeclass Resolution After Apply) and typeclasses eauto, the other being about making the syntax [| ?priority ?pattern] available at Hint/Instance declaration sites, so that one can control the pattern used to filter instances directly (and see which one is inferred when doing Print HintDb) The syntax is backwards compatible, but slightly bad in the Existing Instances (note the "s") case, where one needs to use [(| priority pattern)] to avoid a conflict. I think we could remove that hack too and not support giving explicit priorities and patterns when declaring multiple existing instances. 

It implements this new hint_info as a record of an optional boolean and an optional pattern, the default being no priority and no pattern, in which case these are inferred as usual. The last commit makes the API more compatible in the sense that some functions can take and [hint_info option] where they used to take an [int option], as all the contribs calling this code call it with [None] which should normally get translated to [empty_hint_info]. Anyway comments are welcome both on the doc and the new syntax (the feature is already implemented but can be used for Hint Extern's only right now).

mattam82 commented:
This has been tested here: https://ci.inria.fr/coq/view/coq-contribs/job/coq-contribs/554/console

mattam82 commented:
Not to self: it needs tests and renaming of the first commits

jonleivent commented:
- I do not think it is true that `typeclasses eauto` with no arguments will only directly solve typeclass subgoals.  Consider:

```coq
Variable Foo : Type.
Variable foo : Foo.
Hint Extern 0 Foo => exact foo : typeclass_instances.
Goal Foo.
typeclasses eauto.
Qed.
```

- I don't understand: 

> Using an empty database with unfoldable variables and constants as the first argument of typeclasses eauto hence makes resolution with the local hypotheses use full conversion during unification.

- typo: "typeclasse"

JasonGross commented:
> I don't understand:

Do you not understand what it means, or what the alternative is, or why it works this way?  Which part do you not understand? (I count five separate concepts in that sentence.)

JasonGross commented:
>  I think we could remove that hack too and not support giving explicit priorities and patterns when declaring multiple existing instances.

I think a better replacement would be to allow giving the same priority and pattern to all instances, e.g. `Existing Instances foo bar baz | 3.` would add all of these at priority 3.

JasonGross commented on doc/refman/Classes.tex:
```diff
@@ -382,6 +378,54 @@ \subsection{\tt Context {\binder$_1$ \ldots \binder$_n$}}
+\item When called with no arguments, {\tt typeclasses eauto} uses the
+  {\tt typeclass\_instances} database by default (instead of {\tt core})
+  and will try to solve \emph{only} typeclass goals. Other subgoals are
+  automatically shelved and \emph{must be} resolved entirely when the
```
What's an example of the automatic shelving behavior?  I don't think I've ever seen this happen.

mattam82 replied:
Now it does internally, when the typeclass instance generates a subgoal of non-typeclass conclusion. With my upcoming commits it also shelves the initial subgoals that are not typeclasses.

JasonGross commented on doc/refman/Classes.tex:
```diff
@@ -382,6 +378,54 @@ \subsection{\tt Context {\binder$_1$ \ldots \binder$_n$}}
+  and will try to solve \emph{only} typeclass goals. Other subgoals are
+  automatically shelved and \emph{must be} resolved entirely when the
+  other typeclass subgoals are resolved.
+\item The transparency information of databases is used consistently.
```
Does "consistently" need more explanation?  I'm guessing that it means that for any applied hint, the transparency information in the database the hint comes from is used?

JasonGross commented on doc/refman/Classes.tex:
```diff
@@ -382,6 +378,54 @@ \subsection{\tt Context {\binder$_1$ \ldots \binder$_n$}}
+  
+\item \label{with} {\tt typeclasses eauto with {\ident}$_1$ \ldots {\ident}$_n$}.
+  This variant runs resolution with the given hint databases. It does
+  not treat typeclass subgoals differently than others.
```
The double negative is slightly confusing.  I'd instead write "It treats typeclass subgoals the same as other subgoals."

mattam82 replied:
Right

JasonGross commented on doc/refman/Classes.tex:
```diff
@@ -400,18 +444,105 @@ \subsection{\tt Typeclasses Transparent, Opaque {\ident$_1$ \ldots \ident$_n$}}
+\subsection{\tt Set Typeclasses Module Eta}
+\optindex{Typeclasses Modulo Eta}
+
+This allows eta-conversion during unification of type-classes.
```
Do you want to specify that this is record eta and not function eta?

mattam82 replied:
It is both actually, detailing. Also, this is not working with Filtered Unification on (it always does eta then), I detailed the doc.

JasonGross commented on doc/refman/Classes.tex:
```diff
@@ -400,18 +444,105 @@ \subsection{\tt Typeclasses Transparent, Opaque {\ident$_1$ \ldots \ident$_n$}}
+\subsection{\tt Set Typeclasses Dependency Order}
+\optindex{Typeclasses Dependency Order}
+
+This option (now on by default) respects the dependency order between
```
The "now" is unspecified.  I'd suggest "(on by default since Coq v8.n)".

mattam82 replied:
Right

JasonGross commented on doc/refman/Classes.tex:
```diff
@@ -400,18 +444,105 @@ \subsection{\tt Typeclasses Transparent, Opaque {\ident$_1$ \ldots \ident$_n$}}
+\subsection{\tt Set Typeclasses Limit Intros}
+\optindex{Typeclasses Limit Intros}
+
+This option (on by default in Coq 8.6 and below), controls the ability
```
Grammar says that you can have the comma or the parentheses, but not both.  (If you choose the comma, you need another one after "option".)

mattam82 replied:
Right

JasonGross commented on doc/refman/Classes.tex:
```diff
@@ -400,18 +444,105 @@ \subsection{\tt Typeclasses Transparent, Opaque {\ident$_1$ \ldots \ident$_n$}}
+\optindex{Typeclasses Limit Intros}
+
+This option (on by default in Coq 8.6 and below), controls the ability
+to apply hints while avoiding eta-expansions in the proof term
```
This is function eta, right?  Should this be noted?

mattam82 replied:
Yes, right

JasonGross commented on doc/refman/Classes.tex:
```diff
@@ -400,18 +444,105 @@ \subsection{\tt Typeclasses Transparent, Opaque {\ident$_1$ \ldots \ident$_n$}}
+to apply hints while avoiding eta-expansions in the proof term
+generated. It does so by allowing hints that conclude in an product to
+apply to a goal with a matching product directly, avoiding an
+introduction. \emph{Warning:} can be expensive as it requires rebuilding
```
Consider a complete sentence; add "this" before "can be expensive".

mattam82 replied:
Done

JasonGross commented on doc/refman/Classes.tex:
```diff
@@ -400,18 +444,105 @@ \subsection{\tt Typeclasses Transparent, Opaque {\ident$_1$ \ldots \ident$_n$}}
+\subsection{\tt Set Typeclasses Filtered Unification}
+\optindex{Typeclasses Filtered Unification}
+
+This option, available since Coq 8.6, switches the hint application
```
Perhaps mention the default state of the option.

mattam82 replied:
Done

JasonGross commented on doc/refman/Classes.tex:
```diff
@@ -400,18 +444,105 @@ \subsection{\tt Typeclasses Transparent, Opaque {\ident$_1$ \ldots \ident$_n$}}
+
+This option (on by default) controls the use of typeclass resolution
+when a unification problem cannot be solved during elaboration. 
+With this option on, typeclasses resolution is tried in this case before
```
The use of "before " makes it sound like the alternative is "after" rather than "not at all".  I'd replace "this case before" with "this case, after which"

JasonGross commented on doc/refman/Classes.tex:
```diff
@@ -400,18 +444,105 @@ \subsection{\tt Typeclasses Transparent, Opaque {\ident$_1$ \ldots \ident$_n$}}
+\subsection{\tt Set Typeclasses Strict Resolution}
+\optindex{Typeclasses Strict Resolution}
+
+Type class declarations introduced when this option is set have a
```
Do you mean "Typeclass"? (Mind the spaces)

mattam82 replied:
Yup

JasonGross commented on doc/refman/Classes.tex:
```diff
@@ -400,18 +444,105 @@ \subsection{\tt Typeclasses Transparent, Opaque {\ident$_1$ \ldots \ident$_n$}}
+\optindex{Typeclasses Unique Solutions}
+
+When a typeclasses resolution is launched we ensure that it has a single
+solution or fail. This ensures that the resolution is canonical.
```
This makes resolution much more expensive, right?  Should this be noted?

mattam82 replied:
Yes, it can. Noted

JasonGross commented on doc/refman/Classes.tex:
```diff
@@ -400,18 +444,105 @@ \subsection{\tt Typeclasses Transparent, Opaque {\ident$_1$ \ldots \ident$_n$}}
+\subsection{\tt Set Typeclasses Unique Instances}
+\optindex{Typeclasses Unique Instances}
+
+Type class declarations introduced when this option is set have a
```
"Type class" or "Typeclass"?

mattam82 replied:
Fixed

Zimmi48 commented on doc/refman/Classes.tex:
```diff
@@ -382,6 +378,54 @@ \subsection{\tt Context {\binder$_1$ \ldots \binder$_n$}}
+\item \label{with} {\tt typeclasses eauto with {\ident}$_1$ \ldots {\ident}$_n$}.
+  This variant runs resolution with the given hint databases. It does
+  not treat typeclass subgoals differently than others.
+\end{Variants}
```
Shouldn't the new `dfs eauto` and `bfs eauto` be listed as variants here too?

Zimmi48 commented on doc/refman/Classes.tex:
```diff
@@ -382,6 +378,54 @@ \subsection{\tt Context {\binder$_1$ \ldots \binder$_n$}}
+resolution. This can be used in {\tt Hint Extern}'s for typeclasse
+instances (in hint db {\tt typeclass\_instances}) to
+allow backtracking on the typeclass subgoals created by the lemma
+application, rather than doing type class resolution localy at the hint
```
`localy` -> `locally`

mattam82 replied:
Right

jonleivent commented:
Sorry about the ambiguity.  What I don't understand about:

>  Using an empty database with unfoldable variables and constants as the first argument of typeclasses eauto hence makes resolution with the local hypotheses use full conversion during unification.

- By "empty database" I assume this means one with no Hint Resolves or Hint Externs, is that right?

- Does this empty database have to have some transparency/opaque rules for this statement to apply, or can it be completely devoid of all rules?

- If yes to the above, how do transparency/opaque rules impact resolution with the local hypotheses?

mattam82 commented:
@jonleivent good catch about the "only classes", this was not true at toplevel indeed.
@jason I don't think it makes much sense to allow only the same priority and pattern, I'd rather people use separate Existing Instance commands for now.

mattam82 commented:
@jonleivent About the empty database, I hinted at the fact that you can do "Create HintDb" to make an empty database. I hope this answers your questions. The database has default settings (Print HintDb will tell you). The transparent state has effect on unification (I stated that early on now).

mattam82 commented:
I squashed only some doc changes in the first commit and renamed it. I changed quite a bit of code (most of which is not active by default though) so retesting here:
https://ci.inria.fr/coq/view/coq-contribs/job/coq-contribs/558/

jonleivent commented:
@mattam82 I understand that `typeclasses eauto.` is supposedly different from `Create HintDb emptydb. typeclasses eauto with emptydb.`  I thought the difference mentioned early in the new doc changes is that with the first, non-typeclass subgoals (perhaps only non-top-level?) are shelved and must be solved indirectly, while they are not shelved with the second and can be solved directly.  Is this difference what is meant by "makes resolution with the local hypotheses use full conversion during unification"?

However, I don't even see where non-top-level non-typeclass subgoals get shelved.  Consider this variation on my example from above:

```coq
Variable Foo : Type.
Variable foo : Foo.
Class Bar := {}.
Instance bar1 (f:Foo) : Bar.
Hint Extern 0 Foo => exact foo : typeclass_instances.
Typeclasses eauto := debug.
Goal Bar.
typeclasses eauto.
Qed.
```
which succeeds with debug output:
```
1: apply bar1 on Bar
1.1.1: (*external*) exact foo on Foo
```
The Foo subgoal is certainly not getting shelved, even though it is not at top-level.

I have also noticed that sortclass-typed subgoals are shelved - see [https://coq.inria.fr/bugs/show_bug.cgi?id=5115#c7](url), or this simpler example:
```coq
Class Bar := {}.
Instance bar1 (t:Type) : Bar.
Hint Extern 0 => exact True : typeclass_instances.
Typeclasses eauto := debug.
Goal Bar.
Fail typeclasses eauto.
debug eauto with typeclass_instances.
Qed.
```
Is it just be the sortclass subgoals that have the shelving behavior?

mattam82 commented:
@jonleivent Right, this was a bug, in only_classes mode I was not shelving subgoals of non-class type (Type or otherwise) early enough.

mattam82 commented:
There's a design question I'm not sure about though: should we backtrack if shelved (non-class) subgoals remain after a solution is found or fail globally. I would tend to prefer the second option, as this behavior is meant to provide a way for users to close non-class existentials completely during proof-search and not leave dangling evars (use case: variable maps during reification for example, or some typeclass "output" evar of non-class type that one wants to fill by resolution).

jonleivent commented:
@mattam82 "should we backtrack if shelved (non-class) subgoals remain after a solution is found or fail globally".  Are you asking this about only_classes mode, or in general?

If they are just left on the shelf, one can always achieve the fail with backtracking behavior via `unshelve typeclasses eauto; fail.` and global fail without backtracking with `unshelve once typeclasses eauto; fail.`.  I have a use case for leaving them on the shelf: [https://coq.inria.fr/bugs/show_bug.cgi?id=5115](url), although that probably should be addressed in some other way.

mattam82 commented:
I'm talking about the only_classes case yes.

On Fri, Oct 28, 2016 at 3:53 PM Jonathan Leivent <notifications@github.com>
wrote:

> @mattam82 <https://github.com/mattam82> "should we backtrack if shelved
> (non-class) subgoals remain after a solution is found or fail globally".
> Are you asking this about only_classes mode, or in general?
>
> If they are just left on the shelf, one can always achieve the fail with
> backtracking behavior via unshelve typeclasses eauto; fail. and global
> fail without backtracking with unshelve once typeclasses eauto; fail.. I
> have a use case for leaving them on the shelf:
> https://coq.inria.fr/bugs/show_bug.cgi?id=5115 <http://url>, although
> that probably should be addressed in some other way.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/coq/coq/pull/339#issuecomment-256925700>, or mute the
> thread
> <https://github.com/notifications/unsubscribe-auth/AAGARYyLOk59HZlGVXQkYIVdgzl21xgbks5q4f5egaJpZM4KggO3>
> .
>

mattam82 commented:
Ok, I'm almost there, leaving it for tomorrow.

mattam82 commented:
The semantics of typeclasses eauto (or typeclass resolution run during type inference) is a bit more subtle. What I want is that if resolution succeeds but leaves shelved subgoals, we fail and do not return other potential solutions. But if the resolution succeeds without shelved subgoals, we should be able to backtrack on it (for typeclasses eauto, not during typeclass inference currently). That's what I implemented now. I added a note that "all:once (typeclasses eauto)" corresponds to what happens during type inference now, and will probably move to "all:typeclasses eauto" once the pretyper will be able to backtrack. Comments are welcome! I think I'm done with this for 8.6beta1 otherwise.

mattam82 commented:
Rebased, testing here: 
https://ci.inria.fr/coq/view/coq-contribs/job/coq-contribs/567/

mattam82 commented:
Trying again after a necessary fix in CoRN (silently sheled subgoals)
 https://ci.inria.fr/coq/view/coq-contribs/job/coq-contribs/570/console

mattam82 commented:
The contribs passed with no significant compilation time change.

mattam82 commented:
It misses Set Typeclasses Debug Verbosity n.

mattam82 commented:
TODO cherry pick https://github.com/coq/coq/commit/d74c18826675d398f6c4b1f4e0b30cad7b6da506

mattam82 commented:
All done, let's try the contribs one last time!
https://ci.inria.fr/coq/view/coq-contribs/job/coq-contribs/575/

mattam82 commented:
Ah right, I'm using that a few times, I'll fix it asap
On Mon, 7 Nov 2016 at 09:23, Maxime Dénès <notifications@github.com> wrote:

> *@maximedenes* commented on this pull request.
> ------------------------------
>
> In tactics/hints.ml
> <https://github.com/coq/coq/pull/339#pullrequestreview-7370468>:
>
> > @@ -84,6 +84,8 @@ let secvars_of_hyps hyps =
>    if all then Id.Pred.full (* If the whole section context is available *)
>    else pred
>
> +let empty_hint_info = Vernacexpr.{ hint_priority = None; hint_pattern = None }
>
> This does not compile with OCaml 4.01 (there seems to be other compilation
> errors).
>
> —
> You are receiving this because you were assigned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/coq/coq/pull/339#pullrequestreview-7370468>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAGARdjGuOF8MamidkWK6MByWBXcdLZFks5q7t_kgaJpZM4KggO3>
> .
>

mattam82 commented:
Ah damn, I can't compile with 4.01 on OS X I get linking errors from the native compiler. @maximedenes  can you check? Also I realized that I failed to document two (long-awaited) variants of refine that prevent typeclass resolution from being performed and allow to use refine "safely" in Hints without triggering local resolution on which it is impossible to backtrack afterwards.

mattam82 commented:
(I mean from ocamlopt, nothing to do with native_compute)

herbelin commented:
@mattam82: For the record, that I'm using ocaml 4.01.0 on MacOS 10.11.1 without problems (with camlp5 6.14, ocaml and camlp5 installed with opam).

mattam82 commented:
I'm on 10.10.5, ocaml 4.01.0, ld64-241.9 camlp5 6.15 (it's indeed the cmxs
of ml4 files that fail), everything from opam. I think it's an issue with
XCode's linker (version 6.1.1 of XCode)

On Mon, Nov 7, 2016 at 3:49 PM Hugo Herbelin <notifications@github.com>
wrote:

> @mattam82 <https://github.com/mattam82>: For the record, that I'm using
> ocaml 4.01.0 on MacOS 10.11.1 without problems (with camlp5 6.14, ocaml and
> camlp5 installed with opam).
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/coq/coq/pull/339#issuecomment-258855191>, or mute the
> thread
> <https://github.com/notifications/unsubscribe-auth/AAGARfnY7EobYtSLwDAjxtzTkZ1BJePhks5q7zpygaJpZM4KggO3>
> .
>

jonleivent commented:
It looks like one must do `Typeclasses eauto := debug <depth>` prior to `Set Typeclasses Debug Verbosity <level>`, as doing them in the other order appears to reset the verbosity.  If that is the case, should it be documented?  Or, should the settings be allowed in either order?

Also, considering that the default depth is now unlimited, if one does `Typeclasses eauto := <depth>`, how does one reset the depth to unlimited later?
